exporter/prometheus.go

```diff
@@ -0,0 +1,348 @@
+package exporter
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/manager"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type metric struct {
+	valueType prometheus.ValueType
+	value     float64
+	labels    []string
+}
+
+// PrometheusCollector implements prometheus.Collector
+type PrometheusCollector struct {
+	manager manager.Manager
+
+	errors   prometheus.Gauge
+	lastSeen *prometheus.Desc
+
+	cpuUsageSeconds       *prometheus.Desc
+	cpuUsageSecondsPerCPU *prometheus.Desc
+
+	memoryUsageBytes *prometheus.Desc
+	memoryWorkingSet *prometheus.Desc
+	memoryFailures   *prometheus.Desc
+
+	fsLimit        *prometheus.Desc
+	fsUsage        *prometheus.Desc
+	fsReads        *prometheus.Desc
+	fsReadsSectors *prometheus.Desc
+	fsReadsMerged  *prometheus.Desc
+	fsReadTime     *prometheus.Desc
+
+	fsWrites        *prometheus.Desc
+	fsWritesSectors *prometheus.Desc
+	fsWritesMerged  *prometheus.Desc
+	fsWriteTime     *prometheus.Desc
+
+	fsIoInProgress *prometheus.Desc
+	fsIoTime       *prometheus.Desc
+
+	fsWeightedIoTime *prometheus.Desc
+
+	networkRxBytes   *prometheus.Desc
+	networkRxPackets *prometheus.Desc
+	networkRxErrors  *prometheus.Desc
+	networkRxDropped *prometheus.Desc
+	networkTxBytes   *prometheus.Desc
+	networkTxPackets *prometheus.Desc
+	networkTxErrors  *prometheus.Desc
+	networkTxDropped *prometheus.Desc
+
+	tasks *prometheus.Desc
+
+	descs []*prometheus.Desc
+}
+
+// NewPrometheusCollector returns a new PrometheusCollector
+func NewPrometheusCollector(manager manager.Manager) *PrometheusCollector {
+	c := &PrometheusCollector{
+		manager: manager,
+		errors: prometheus.NewGauge(prometheus.GaugeOpts{
+			Namespace: "container",
+			Name:      "scrape_error",
+			Help:      "1 if there was an error while getting container metrics, 0 otherwise",
+		}),
+		lastSeen: prometheus.NewDesc(
+			"container_last_seen",
```

Comment:
The namespace is only one of the options in `...Opts`, which are used to construct fully fledged metric objects for Prometheus. Those are meant to be used for "native" Prometheus metrics. What we are doing here is copying existing metrics into Prometheus metrics. For that, we create some of the Prometheus metrics on the fly during scraping, which does not follow the 'high level' path with the Opts object, but assembles the necessary parts one level further down (in this case a metric descriptor is created, which doesn't feature setting a namespace via Opts - so the "container" part has to be included in the name).